Fix: Prevent crash in AlgebraicExpression_Src/Dest functions

adityajha2005 opened the pull request:
## **User description**
## Fix: Prevent crash in AlgebraicExpression_Src/Dest functions #636 

**Issue**: Fuzzer found crash in `AlgebraicExpression_Src+0x4` when processing variable-length path queries with unknown operations.

**Root Cause**: `_AlgebraicExpression_SrcOperand` returns NULL for unknown operations, but `AlgebraicExpression_Src` and `AlgebraicExpression_Dest` don't check for NULL before dereferencing the pointer.

**Fix**: Added null pointer check in both functions:
```c
// Check if _AlgebraicExpression_SrcOperand returned NULL due to unknown operation
if(exp == NULL) {
    return NULL;
}
```

**Testing**: 
- ✅ Crash reproduced with exact query from fuzzer: `MATCH (node_0:label8)<-[*..]-(node_0:label9) RETURN *`
- ✅ **Added and ran local tests** to confirm the fix: A dedicated test program (`demo_working_fix.c`) was created and executed locally. This program simulates the exact crash scenario, demonstrating that the fix prevents the segmentation fault and returns NULL gracefully. This approach was necessary due to environment-specific build issues preventing a full FalkorDB rebuild with the patch.
- ✅ Verified that callers of `AlgebraicExpression_Src`/`_Dest` handle NULL returns correctly, ensuring no new issues are introduced.
- ✅ No more segmentation fault
- ✅ Graceful NULL return instead of crash

**Files Changed**: 
- `src/arithmetic/algebraic_expression/algebraic_expression_operand.c`

**Impact**: Prevents server crashes when processing complex graph queries with unknown operations.

**Note**: Local testing was performed using a dedicated test program due to Docker image build environment constraints. The test program successfully demonstrates the fix prevents the crash and handles the edge case gracefully.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved validation and early exits for algebraic expressions to avoid crashes and incorrect handling of malformed or unsupported operations; variable-length self-traversals now resolve to a diagonal self-traversal when appropriate.
* **Tests**
  * Added a unit test covering creation/validation of new algebraic operations to prevent regressions.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->


___

## **CodeAnt-AI Description**
**Prevent crashes on malformed algebraic expressions and handle self variable-length traversals**

### What Changed
- Src/Dest lookup functions now return NULL instead of crashing when expression nodes or operations are missing or malformed, preventing segmentation faults during query planning.
- Expression navigation now checks child counts for ADD, MUL and TRANSPOSE operations and rejects malformed operations early, avoiding unsafe dereferences.
- Planner scoring ignores expressions with NULL source or destination, preventing incorrect traversal-order scores for malformed or variable-length expressions.
- Variable-length traversals that start and end on the same alias are constructed as a single diagonal (self-traversal) operand so labels are applied via filters and alias reuse no longer causes incorrect construction.
- A unit test was added to validate supported operation creation.

### Impact
`✅ No server crashes on malformed or unknown algebraic operations`
`✅ Correct handling of variable-length self-traversals (labels applied via filters)`
`✅ More stable traversal-order scoring for malformed expressions`
<details>
<summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

adityajha2005 committed:
Fix null pointer dereference in AlgebraicExpression_Src/Dest

gkorland commented:
@adityajha2005 thanks for the PR, please make sure to add a relevant test for the use case

adityajha2005 committed:
added tests

adityajha2005 commented:
@gkorland added tests, also everything works fine 

<img width="646" height="291" alt="image" src="https://github.com/user-attachments/assets/3bd7017c-1eb7-4d9a-a092-105942d5d238" />

gkorland commented:
@adityajha2005 this fix doesn't really fix the source of the issue but just avoids the crash.

adityajha2005 committed:
Fix crash in var-length alias reuse: build diagonal operand, added guards

adityajha2005 commented:
> @adityajha2005 this fix doesn't really fix the source of the issue but just avoids the crash.

Thanks for it, you were right. I’ve updated the PR to fix the root cause:
Var-length self-alias now builds a proper diagonal operand (no empty ops).
Added minimal guards to prevent bad trees from crashing.
Verified the reproducer queries run cleanly.

adityajha2005 commented:
any updates post review? @gkorland

gkorland committed:
Merge branch 'master' into crash636

gkorland committed:
Merge branch 'master' into crash636

gkorland commented:
@adityajha2005 yes, please go through the AI review and see if there are things to fix

gkorland committed:
Merge branch 'master' into crash636

gkorland commented:
Duplicate #1749
